feature: todolist- C.R.U.D, signin, signup 개발 완료

201steve opened the pull request:
## 개발한 화면
| title  | view |
|--------|--------|
| createTodo |![createTodo](https://user-images.githubusercontent.com/79638133/235801274-99a3b5dd-3fa0-447c-8be2-40580cf080c8.gif)|
| display detail |![displayDetail](https://user-images.githubusercontent.com/79638133/235801404-3d267ef3-14a3-44ba-919c-a44f83bb0931.gif)|
| delete todo | ![deleteTodo](https://user-images.githubusercontent.com/79638133/235801416-5be7d239-2ad7-4b2b-9d4e-383aadfa6c59.gif) |
| updateTodo | ![update](https://user-images.githubusercontent.com/79638133/235801454-009bf3a8-0ffa-44e0-b35f-ac961baea5ac.gif)|
| signup |![signup](https://user-images.githubusercontent.com/79638133/235801476-c34f2835-2d4e-4020-b2fd-468e5dff9d77.gif)| 
| login |![signin](https://user-images.githubusercontent.com/79638133/235801500-ae75f154-ce95-48d9-a3ad-12b56dea4266.gif)| 
| refresh but steady| ![refreshbutsteady](https://user-images.githubusercontent.com/79638133/235801820-3e084ad6-059c-40c2-b472-1e8add8a1662.gif)| 


## 개발 내용
| commit  | Detail |
|--------|--------|
|[48fdafe](https://github.com/201steve/todo/pull/8/commits/48fdafe4f05839682faa40867c3ac4c35bcaea37)| layout signup, signin, add login validate,토큰 여부에 따른 route 처리 추가|
|[3095ef7](https://github.com/201steve/todo/pull/8/commits/3095ef7e8461c26be34b27ab35e3b9dd54844b49)|api주소 관리용 파일 추가, 토큰 유효하지않다면 로그인페이지로 리다이렉트 추가|
| [d5baa95](https://github.com/201steve/todo/pull/8/commits/d5baa9549fc9af43d99677a2cb6e3a196a70c0b6) | 특정 todo를 선택하면, 상세내역 출력 |
| [5216eca](https://github.com/201steve/todo/pull/8/commits/5216eca39996e3ff7ef41165ad8d1599cf379a26) | label태그 삭제, 체크박스에 체크 되야 detail이 보이던 방식에서 todo 클릭 하면 보이는것으로 수정, detail 닫기 버튼 추가 ,title의길이가 특정 index이상 되면 ... 으로 collapse되게 view 수정, input scale transition 삭제 |
| [34eecbf](https://github.com/201steve/todo/pull/8/commits/34eecbf49b4ba5d9c6cbb08681dcf4fde868a9ce) | create todo 기능 추가| 
| [5229454](https://github.com/201steve/todo/pull/8/commits/52294540ecbd2d5b1ee2bb2cafa2e1e6e6e0f0f0) | delete 버튼을 누르면 todo 삭제 되는 기능 추가 ,delete 버튼을 누르고 삭제되면 re-render되도록 setState 구성 ,createTodo 하면re-render 되도록setState수정, update 용도 modal 구현중 | 
|5b96eb3233c2e50a5ec92905bef85e67044573ff&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| todo update 기능 추가 | 
|536dbc77e812f2f3f759018c37aadf52dcd83487 | 새로고침해도 유지되도록 기능 추가 |
|https://github.com/201steve/todo/pull/9/commits/e76c6a58b03c967e7811f62490026412a5dcd51b|axios로 전환, 불필요 state 삭제, detail 여부에 따라 조건부 렌더링으로 변경 |
|https://github.com/201steve/todo/pull/9/commits/e04b37dd52e0acd73a7da3d9d9ce6c2e11ccd947| service code 분리, 로그인 여부에 따른 privateRouter 설정, axios interceptors추가로 instance request 전 header에 Authorization token update  |
## Assignment 1 - Login / SignUp

- /auth 경로에 로그인 / 회원가입 기능을 개발합니다
  - 로그인, 회원가입을 별도의 경로로 분리해도 무방합니다
  - [x] 최소한 이메일, 비밀번호 input, 제출 button을 갖도록 구성해주세요
- 이메일과 비밀번호의 유효성을 확인합니다
  - [x] 이메일 조건 : 최소 `@`, `.` 포함
  - [x] 비밀번호 조건 : 8자 이상 입력
  - [x] 이메일과 비밀번호가 모두 입력되어 있고, 조건을 만족해야 제출 버튼이 활성화 되도록 해주세요
- 로그인 API를 호출하고, 올바른 응답을 받았을 때 루트 경로로 이동시켜주세요
  - [x] 응답으로 받은 토큰은 로컬 스토리지에 저장해주세요
  - [x] 다음 번에 로그인 시 토큰이 존재한다면 루트 경로로 리다이렉트 시켜주세요
  - [x] 어떤 경우든 토큰이 유효하지 않다면 사용자에게 알리고 로그인 페이지로 리다이렉트 시켜주세요

## Assignment 2 - Todo List

- Todo List API를 호출하여 Todo List CRUD 기능을 구현해주세요
  - [x] 목록 / 상세 영역으로 나누어 구현해주세요
  - [x] Todo 목록을 볼 수 있습니다.
  - [x] Todo 추가 버튼을 클릭하면 할 일이 추가 됩니다.
  - [x] Todo 수정 버튼을 클릭하면 수정 모드를 활성화하고, 수정 내용을 제출하거나 취소할 수 있습니다.
  - [x] Todo 삭제 버튼을 클릭하면 해당 Todo를 삭제할 수 있습니다.
- 한 화면 내에서 Todo List와 개별 Todo의 상세를 확인할 수 있도록 해주세요.
  - [x] 새로고침을 했을 때 현재 상태가 유지되어야 합니다.
  - [x] 개별 Todo를 조회 순서에 따라 페이지 뒤로가기를 통하여 조회할 수 있도록 해주세요.
- 한 페이지 내에서 새로고침 없이 데이터가 정합성을 갖추도록 구현해주세요

  - [x] 수정되는 Todo의 내용이 목록에서도 실시간으로 반영되어야 합니다
## 회고

- create와 delete때 어떻게 re-render시킬까에 대한 고민이 많았는데, 차근차근 해결.
- delete할때 navigate로 넘기고 params로 받아오면 이전 id가 url에 남아있어서 2번 클릭해야 삭제되는 버그가 있었는데 선택한 element의 id를 받아와서 api로 보내주는 방식으로 수정.
- detail을 열어놓은상태로 목록의 todo를 다 지워도 delete가 그대로 남아있어서 useEffect로 todoList의 length가 0 이면 false되도록 일단 땜질.
- update 용도로 만드는 modal을 portal로 쓸지, absolute로 띄울지 여러 시도해보는중.
- 그러다가 배경색 opacity를 적용하는데서 꽤나 시간을 잡아먹었고, 정말 어이없이 간단하게도 background-color의 rgba 부분의 값을 부여하면 해결되는 문제였음.
-update 감이 잘 안잡혔는데, 해결했다. CRUD 는 끝났고, 정합성, 다른 기능들을 좀 써보고 리팩토링도 필요하다. fetch 코드 반복이 많은데 그것부터 해봐야겟다.
- axios 처음써봤는데, 반복되는 옵션들을 instance 생성해서 쓸 수 있는게 굉장히 편한 것 같다. JSON파싱도 해줘서 비동기함수로 편하게 쓸 수 있는게 장점같다. 왜 진즉 써볼 생각을 안했을까.. 너무 좋다
- 이전까지는 특정페이지에 접근했을 때 로그인 여부를 판단해서 일단 useEffect가 돌기 전에 렌더부터된 화면이 노출됐었는데, PrivateRouter로 경로 설정하는걸 알게되서, 애시당초 로그인 여부를 Router에서 판단해 진입조차 안되도록 수정.
- 로그아웃된상태에서 로그인을하면 리스트로 보내지도록 구성했는데, 리스트로와서 꼭 새로고침 한번을 해야 했다. 원인은 localstorage에 값이 초기엔 null이라 unAuthorization 인것. 그래서 리스트 컴포넌트가 마운트되고-> JSX render->get 요청에서 get요청이 안됐던것. 해결은, axios에 interceptors라는 방법으로 해결. interceptors가 하는일은 api request 하기전, response 받기 전에 특정 동작을 수행할 수 있는데, list페이지로 넘어와서 get 요청하기전에 interceptors가 낚아채서 headers에 있는 Authorization 값을 null에서 JWT로 갈아끼우고 api call 하는 순서로 진행되어 이상없이 list에 todoList 렌더..
- 다른 api call은 다 잘됐는데, 유난히 signup할때만 안되더라. 그래서 인스턴스에 config를 봤더니, baseUrl로 되어있던것...ㅠㅠㅠㅠㅠ4시간날림...`baseURL`이다..

201steve committed:
add: list 페이지 레이아웃 & fetching test
- add: todo hover transition

201steve committed:
add:auth
- layout signup, signin
- add login validate
- 토큰 여부에 따른 route 처리 추가

201steve committed:
add: 기능추가
- 특정 todo를 선택하면, 상세내역 출력

201steve committed:
auth 기능추가
- api주소 관리용 파일 추가
- 토큰 유효하지않다면 로그인페이지로 리다이렉트 추가

201steve committed:
mofidy:detail view 수정
- label태그 삭제
- 체크박스에 체크 되야 detail이 보이던 방식에서 todo 클릭 하면 보이는것으로 수정
- detail 닫기 버튼 추가
- title의길이가 특정 index이상 되면 ... 으로 collapse되게 view 수정
- input scale transition 삭제

201steve committed:
create todo 기능 추가

201steve committed:
Update README.md

201steve committed:
ADD: delete기능추가와 view수정
- delete 버튼을 누르면 todo 삭제 되는 기능 추가
- delete 버튼을 누르고 삭제되면 re-render되도록 setState 구성
- createTodo 하면re-render 되도록 setState수정
- update 용도 modal 구현중

201steve committed:
add:upate todo 추가

YunKukPark commented on src/Router.js:
```diff
@@ -1,14 +1,17 @@
+
+import List from "./pages/list/List";
+import Auth from "./pages/Auth/Auth";
+import Signin from "./pages/Auth/Signin";
```
이거 jsconfig.json으로 절대경로 셋팅하면 더 좋을 것 같습니다!

https://dubaiyu.tistory.com/271

201steve replied:
추가하여서 절대경로로 돌렸읍니다..👍🏻

YunKukPark commented on src/api/api.js:
```diff
@@ -0,0 +1,11 @@
+  updateTodo: `${BASE}/todos/`,
+  deleteTodo: `${BASE}/todos/`,
+  signIn: `${BASE}/users/login`,
+  signUp: `${BASE}/users/create`,
```
how about
```suggestion
  getTodos: `${BASE}/todos`,
  getTodoById: (id) => `${BASE}/todos/${id}`,
  createTodo: `${BASE}/todos`,
  updateTodo: (id) => `${BASE}/todos/${id}`,
  deleteTodo: (id) => `${BASE}/todos/${id}`,
  signIn: `${BASE}/users/login`,
  signUp: `${BASE}/users/create`,
```

YunKukPark replied:
만약 axios를 사용한다면 api Instance를 만들고 (ex. AuthApi, TodoApi)
Authorization이 필요한 부분에 자동으로 주입할 수도 있어요!

```ts
import axios from 'axios';

const baseUrl = `ㅓ쩌고 저쩌고`;

export const TodoApi = axios.create({ baseURL: `${baseUrl}/todos` });
TodoApi.defaults.headers.common[
  'Authorization'
] = `Bearer ${localStorage.getItem('accessToken')}`;

export const AuthApi = axios.create({ baseURL: `${baseUrl}/auth` });
```

YunKukPark replied:
실제 페이지단에서는 이런식으로

```tsx
  const fetchTodos = async () => {
    const res = await TodoApi.get('');
    setTodos(res.data);
    setIsLoading(false);
  };

  const createTodo = async (userInput: string) => {
    if (!userInput) return;
    const res = await TodoApi.post('', { todo: userInput });
    setTodos([...todos, res.data]);
  };

  const deleteTodo = async (id: number) => {
    await TodoApi.delete(`/${id}`);
    setTodos(todos.filter(todo => todo.id !== id));
  };

  const updateTodo = async (payload: {
    id: number;
    todo?: string;
    isCompleted?: boolean;
  }) => {
    const { id, todo, isCompleted } = payload;
    const res = await TodoApi.put(`/${id}`, { todo, isCompleted });
    setTodos(todos.map(todo => (todo.id === id ? res.data : todo)));
  };
```

YunKukPark commented on src/pages/Auth/Auth.jsx:
```diff
@@ -0,0 +1,73 @@
+          value={email}
+          name="email"
+          onChange={getUserInfo}
+          className={`p-3 px-4 border-2 w-[300px] ${
+            email.includes("@") && "border-emerald-200"
+          } rounded-md mb-5 focus:outline-none `}
```
className에 너무 많은 관심이 쏠려있는 것 같습니다!!
이렇게 되면 rule에 대한 추가 요구사항이 들어왔을 때에 핸들링이 어려워 질 것 같아요.

rule에 대한 내용은 따로 util이나, 해당 컴포넌트 바깥에서 관리 하는건 어떨까요??

like this

1. 단순 분리
```ts
// 바깥쪽
const RULES = {
  id: (value) => value.includes('@'),
  password: (value) => value.length >= 8,
}
```

```suggestion
          className={`p-3 px-4 border-2 w-[300px] ${
            RULES.id(email) && "border-emerald-200"
          } rounded-md mb-5 focus:outline-none `}
```

2. Input Component 분리
그런데 작성하다보니 이 input은 사용자가 작성한 내용에 따라 보여지는 view가 달라집니다.
그래서 Core Input을 따로 만들고, LabeldInput 으로 Higher Ordering 하면 좋을 것 같아요!!
위에 대한 예시는 코드샌드박스 남겨놓습니다!!

https://codesandbox.io/s/input-jal-darugi-rwryx8?file=/src/App.js:0-34

YunKukPark replied:
이렇게 하면 어떤 이점도 있냐면, 
input에 관련된 모든 RULE이 여기에 정리되어 있으니
1. 전체가 validate되었는지 한줄로 확인 할 수 있다.
```jsx
                                                   __👇함수__(함수호출)
Object.entries(userInfo).every(([key, value]) => RULES[key](value))
                                                       ¯¯¯   ¯¯¯¯¯
                              userInfo의 키값👆        👆Rules의 value로 넣을 값
```

2. 상황에 따라 filtering 할 수 있다 => RULE 확인을 안할 수 있다
```jsx
Object.entries(userInfo)
  .filter(필터로직)
  .every(확인로직)
```

201steve replied:
요거슨 auth 브랜치에서 해결해버렸습니다요

saengmotmi replied:
전 갠적으로 react-hook-form에 yup 같은거 물려서 쓰는게 맞다고 봅니다 ㅋㅋㅋ

201steve replied:
react-hook-form, yup 한번 써보겠습니다..

YunKukPark commented on src/pages/Auth/Signin.jsx:
```diff
@@ -0,0 +1,91 @@
+import React, { useState } from "react";
+import { useNavigate, Link } from "react-router-dom";
+
+const Signin = () => {
```
여기는 Auth 쪽 리뷰 확인하시면 좋을 것 같습니다!

201steve replied:
여기도 auth브랜치에서 해-결

YunKukPark commented on src/pages/list/List.jsx:
```diff
@@ -0,0 +1,122 @@
+import TodoModal from "./TodoModal/TodoModal";
+
+const List = () => {
+  const [todoList, setTodoList] = useState([]);
+  const [isTodoSelected, setIsTodoSelected] = useState(false);
+  const [selectedTodo, setSelectedTodo] = useState({});
```
같은 todo에 대한 핸들링이라서 object형식으로 가져가면 좋을 것 같은데 어떻게 생각하시는지요!!

아래는 간단하게 만들어본 Todo의 타입인데 정일님의 todo 비즈니스로직에 맞춰 타입은 수정해주세요!

1. todo의 타입 정의
```tsx
type TodoType = {
  id: number;
  title: string;
  detail: string; 
  ...키타 등등
  isSelected: boolean;
  isCompleted: boolean;
};
```

2. todo의 state는 굉장히 많은 action들을 가지고 있는 것 같습니다.
시간이 되신다면 useReducer를 사용하여 action들을 모아놓고 관리하면 좋을 것 같습니다!

201steve replied:
isTodoSelected는 불필요해져서 삭제해버렸습니다요 해-결

YunKukPark commented on src/pages/list/List.jsx:
```diff
@@ -0,0 +1,122 @@
+          id={id}
+          title={title}
+          content={content}
+          todoList={todoList}
+          whichISelected={whichISelected}
+          displayDetail={displayDetail}
+          deleteTodo={deleteTodo}
```
위에서 todo의 관심을 갖는 녀석들을 모두 묶었으니 이제 넘겨주는 prop도 간소화 되겠군요~~

201steve replied:
한층 간소화되었습니다..

YunKukPark commented on src/pages/list/Todo.jsx:
```diff
@@ -0,0 +1,126 @@
+
+  const submitEditedTodo = async (e) => {
+    e.preventDefault();
+    const response = await fetch(`${api.updateTodo}${id}`, {
+      method: "PUT",
+      headers: {
+        "Content-type": "application/json",
+        Authorization: localStorage.getItem("todo-token"),
+      },
+      body: JSON.stringify({ title: editedTodo.title, content: editedTodo.content }),
```
api.js 의 Instance만들 때 바로 auth 붙혀주는거 보면 좋을 것 같습니다!

201steve replied:
axios로 해-결

YunKukPark reviewed (commented):
🐝 beeeeee~~~

201steve committed:
refactor:불필요한 state 삭제, id가 필요한 api 주소 함수표현식으로 수정
- todo,todoDetail 컴포넌트 분리
- 새로고침시 id 남아있지않도록 path parameter 초기화

201steve committed:
useReducer로 리팩토링중

201steve committed:
refactor: axios 도입, 불필요한 state 제거
- api에는 auth와 todolist instance 생성, api 관리
- client에는 인스턴스의 config 관리로 나눔
- pathParameter 여부로 detail 조건부 렌더로 수정

saengmotmi commented on src/api/client.js:
```diff
@@ -0,0 +1,10 @@
+import { api, authApi, todoApi } from "./api";
+
+export const client = {
+  getTodos: () => todoApi.get(api.getTodos),
+  createTodo: (title, content) => todoApi.post(api.createTodo, { title, content }),
+  updateTodo: (id, title, content) => todoApi.put(api.updateTodo(id), { title, content }),
+  deleteTodo: (id) => todoApi.delete(api.deleteTodo(id)),
+  signIn: (email, password) => authApi.post(api.signIn, email, password),
+  signup: (email, password) => authApi.post(api.signUp, email, password),
+};
```
@201steve 
https://github.com/feedoong/feedoong-frontend/blob/dev/next-app/src/services/subscriptions/index.ts 이런 느낌으로 정리하는 건 어떨까요? 저는 api call을 하는 함수 그 자체는 services 폴더 아래에 엔티티별로 분류해놓는 편이에요.

그건 그거고 이 코드에서 좀 의문이 드는 지점이 몇 군데 있어요.
- api/client가 뜻하는 바는 무엇인가요? client가 아닌 파일도 있을 수 있는거 처럼 느껴져요
- client라는 이름이 todos / auth 를 묶어 표현할 수 있는 개념인지 잘 모르겠어요
- api라는 객체는 url 상수 값들을 모아놓고 있는 셈인데 api라는 이름이 붙는 것이 가장 정확한 표현인지 잘 모르겠어요

201steve replied:
캄사합니다 api call하는 함수는 services 폴더에서 각각 나눠봤습니다!
이 파일도 삭제 했어요!

saengmotmi commented on src/hooks/useReducer.js:
```diff
@@ -0,0 +1,10 @@
+import { useReducer } from "react";
+
+const todoReducer = (state, action) => {
```
작성 중인 로직인가요?

201steve replied:
아 요건 지워야했는데 놓쳤네요..지웠습니다 :-)

saengmotmi commented on src/pages/Auth/Auth.jsx:
```diff
@@ -0,0 +1,73 @@
+          } rounded-md mb-5 focus:outline-none`}
+        />
+        <button
+          className="disabled:bg-red-100 enabled:bg-emerald-100 enabled:active:bg-emerald-300 rounded-md "
```
테일윈드는 아무리 봐도 정 붙이기는 어렵네요 ㅋㅋㅋㅋ

201steve replied:
저도...정이 잘 안붙더라구요 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

saengmotmi commented on src/api/api.js:
```diff
@@ -0,0 +1,26 @@
+
+export const todoApi = axios.create({
+  baseURL: BASE,
+  headers: { "Content-Type": DEFAULT_HEADER, Authorization: AUTH_HEADER },
```
localStorage에서 값을 찾지 못하면 null이 들어가지 않을까요?

201steve replied:
방법을 찾았습니다..`todoService.interceptors.request.use()`로 todoService request전에 낚아채서 Authorization을 업데이트하면 되네요 ㅠㅠㅠ

saengmotmi commented on src/api/api.js:
```diff
@@ -0,0 +1,26 @@
+const AUTH_HEADER = localStorage.getItem("todo-token");
+const DEFAULT_HEADER = "application/json;charset=UTF-8";
+
+export const api = {
```
이미 axios instance에서 base url 세팅을 해주고도 있고요... 음 사실 이 부분은 api/client랑 그냥 함수 하나로 합쳐져도 되는 코드가 아닌가 싶어요. 리뷰 드린 내용처럼요.

201steve replied:
오!! axios를 이제 막 써봐서 어떻게 써야 할지 잘 몰랐습니다
하나로 합치니 좋군요..감사합니다

saengmotmi commented on src/pages/Auth/Auth.jsx:
```diff
@@ -0,0 +1,73 @@
+  const validation = email.includes("@") && password.length >= 8;
+
+  const signUp = async () => {
+    const response = await fetch(api.signUp, {
```
axios 사용하는 곳과 fetch 사용하는 곳의 차이가 있나요?

201steve replied:
auth 코드는 auth브랜치에서 손보려고 했는데, 커밋이 합쳐져서
fetch로 되어있는건 다 axios로 수정 했습니다 :-)

saengmotmi commented on src/pages/Auth/Signin.jsx:
```diff
@@ -0,0 +1,91 @@
+
+  const signUp = async (e) => {
+    e.preventDefault();
+    const response = await fetch("http://localhost:8080/users/login", {
```
api 호출이 실패했을 경우엔 어떻게 되려나요?

201steve replied:
catch문으로 alert을 띄우는 방법으로 수정했습니다.
이거말고 다른 대안이 있을까요...?ㅠㅠ

saengmotmi commented on src/pages/list/List.jsx:
```diff
@@ -0,0 +1,80 @@
+
+  const deleteTodo = async (e, selectedId) => {
+    const response = await client.deleteTodo(selectedId);
+    if (response.statusText === "OK") {
```
이 if문이 있어야만 호출 성공을 판단할 수 있나요?

201steve replied:
try catch로 수정해서 try에 일단 호출성공했다는 가정하에 다음 동작들을 정의했습니다.
try에서 만약 호출이 안됐다면 catch에서 확인하는 방식으로 수정했어요!

그런데 이게 맞나요...?

saengmotmi commented on src/pages/list/List.jsx:
```diff
@@ -0,0 +1,80 @@
+      navigate("/signin");
+    }
+    getTodo();
+  }, []);
```
useEffect에서 로직을 판별하면
렌더링 -> useEffect -> 조건문 -> 라우팅 순서로 작동할텐데 그럼 일단 화면이 보여버리지 않을까요?
권한이 없다면 애초에 이 페이지로 들어오면 안될 거 같은데...

201steve replied:
조건부로 priavateRoute 설정하는거로다가 적용했습니다!!
하나 또 배웠습니다 크으...👍🏻

saengmotmi commented on src/pages/list/TodoModal/TodoModal.jsx:
```diff
@@ -0,0 +1,58 @@
+      >
+        <input
+          onChange={getNewTodo}
+          ref={modalRef}
```
input에 걸리는건데 modalRef는 이름이 좀 부적절한거 같고요 autoFocus라는 속성이 더 적절하지 않을까요?

201steve replied:
아! 그렇게도 이름지을 수 있다니! 감사합니다

saengmotmi commented on src/pages/list/TodoModal/TodoModal.jsx:
```diff
@@ -0,0 +1,58 @@
+    e.preventDefault();
+    const response = await client.createTodo(title, content);
+    alert("todo가 생성되었습니다");
+    setTodoModalIsOpend((prev) => !prev);
```
이런 함수는 어떤 컴포넌트에서는 toggleFoo 형태로 쓰시던데 일관성이 있으면 좋을거 같습니다

201steve replied:
네!! 함수를 만들어놓고 setState를 또 가져다 썻네요..ㅠ

saengmotmi commented on src/pages/list/Todo/Todo.jsx:
```diff
@@ -0,0 +1,113 @@
+      const editedIndex = prevTodoList.findIndex(({ id }) => id === updatedTodo.id);
+      if (editedIndex !== -1) {
+        const updatedTodoList = [...prevTodoList];
+        updatedTodoList[editedIndex] = updatedTodo;
```
```suggestion
    setTodoList((prevTodoList) => {
      const isTodoEdited = prevTodoList.find(({ id }) => id === updatedTodo.id);
      if (isTodoEdited) {
        const updatedTodoList = [...prevTodoList];
        updatedTodoList[editedIndex] = updatedTodo;
```
`-1` 자체로는 어떠한 의미를 표현하지 못하기 때문에(매직 리터럴) 한번 더 생각해야 하는 반면, 그냥 find를 쓰면 되지 않나 싶기도 하고.. 그리고 `-1`이 `아닌 것`은? 이라는 로직은 뭔가 묘하게 이중부정 같기도 하네요 그냥 `찾으면` 이라는 로직이 자연스러워 보입니다

201steve replied:
find로 수정했습니다!
index찾아서 갈아껴야지~ 생각했는데 
find가 바로 해당 요소를 반환하니까 보기 편하네요..  감사합니다👍🏻

saengmotmi commented on src/pages/Auth/Signin.jsx:
```diff
@@ -0,0 +1,91 @@
+        <button
+          className="disabled:bg-red-100 enabled:bg-emerald-100 enabled:active:bg-emerald-300 rounded-md "
+          type="submit"
+          disabled={!validation}
```
뭐에 대한 유효성 검증인지 알기가 힘든데요. isSubmitEnabled 이런 느낌은 어떨까요?

201steve replied:
아..이거 왜이런 이름으로 변수를 지었는지..뭔가 힘들었나봅니다...수정했어요! :-)

201steve committed:
refactor : api call 함수 분리, 로그인여부 처리
- api.js에서 인스턴스랑 service 분리
- useEffect에서 토큰 확인여부에서 pprivateRouter로 변경해 로그인이 안되면 진입조차 안되게 수정
- api.js에서 로그인 후 인스턴스 request전에 token 업데이트 interceptors 추가
- fetch를 전부 axios로 수정

201steve reviewed (commented):
감사합니다 종택님 많이 배웠습니다!!

201steve reviewed (commented):
@saengmotmi 감사합니다 종택님 많이 배웠습니다!!
